keps/sig-apps/20190715-rollingUpdate-enhancement-for-daemonset.md

```diff
@@ -0,0 +1,182 @@
+
+1. When rollingUpdate, user should have some chioces to select some specific or random nodes to update pods running on it.
+
+### Goals
+
+- Add field `Paused` for `DaemonSetSpec`.
+- Add new fields `Selector` and `Partition` for existing RollingUpdateDaemonSet.
+
+### Non-Goals
+- This implement is only implemeted to affect to update strategy of Daemonset. No other behaviors of Daemonset will be changed.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+There are serveral defects of current community design. Such as:
+* Pause and Resume. DaemonSet does not has the fetature of Pause/Resume in update。 Especially in large or huge scale cluster, in order to make the cluster high avaiable and certainly correct image version, Kubernetes should server the ability that user can pause DaemonSet update process and check if target version image meet their expect. If users or administrators' test verified new version pass probe test, they can resume DaemonSet update.
+
+* Batch update. As for a mature system, gray update is a necessary ability. However there is no any enhancement on this area. Now the community version do not supports any partion process in update like other workload, such StatefulSet.
+
+### Implementation Details
+
+#### API Changes
+
+Following changes will be made to the Rolling Update Strategy for StatefulSet.
+
+```go
+// DaemonSetSpec is the specification of a daemon set.
+type DaemonSetSpec struct {
+  ...
+    // Indicates that the daemonset is paused and will not be processed by the
+  // daemonset controller.
+  // +optional
+  Paused bool
+  ...
+}
+```
+
+```go
+type RollingUpdateDaemonSet struct {
+	// The maximum number of DaemonSet pods that can be unavailable during the
+	// update. Value can be an absolute number (ex: 5) or a percentage of total
+	// number of DaemonSet pods at the start of the update (ex: 10%). Absolute
+	// number is calculated from percentage by rounding up.
+	// This cannot be 0.
+	// Default value is 1.
+	// Example: when this is set to 30%, at most 30% of the total number of nodes
+	// that should be running the daemon pod (i.e. status.desiredNumberScheduled)
+	// can have their pods stopped for an update at any given
+	// time. The update starts by stopping at most 30% of those DaemonSet pods
+	// and then brings up new DaemonSet pods in their place. Once the new pods
+	// are available, it then proceeds onto other DaemonSet pods, thus ensuring
+	// that at least 70% of original number of DaemonSet pods are available at
+	// all times during the update.
+  // +optional
+  MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,1,opt,name=maxUnavailable"`
+
+	// A label query over nodes that are managed by the daemon set RollingUpdate.
+	// Must match in order to be controlled.
+	// It must match the node's labels.
+	Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,2,opt,name=selector"`
```

Comment:
Would you explain more about how this `selector` works with rolling update?

Note that DaemonSet works with node selector. 
https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#running-pods-on-only-some-nodes